File: app/src/main/java/com/example/homeworkproject/viewmodels/LocationViewModel.java
package com.example.homeworkproject.viewmodels;

import static com.example.homeworkproject.model.ApiState.Status.ERROR;
import static com.example.homeworkproject.model.ApiState.Status.LOADING;
import static com.example.homeworkproject.model.ApiState.Status.SUCCESS;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;

import com.example.homeworkproject.client.Api;
import com.example.homeworkproject.model.ApiState;
import com.example.homeworkproject.model.Country;
import com.example.homeworkproject.model.Province;
import com.example.homeworkproject.repository.Repository;

import java.util.ArrayList;

public class LocationViewModel extends ViewModel {

    private final Repository repository;
    public final MutableLiveData<ApiState.Status> currentState = new MutableLiveData<>();
    public MutableLiveData<ArrayList<Province>> provinceData;
    public MutableLiveData<ArrayList<Country>> countryData;
    public String currentVal;

    public LocationViewModel(){
        this.repository = new Repository();
    }

    public void getLiveCountryData() {

        if (countryData ==null || countryData.getValue() == null){
            currentState.setValue(LOADING);
            countryData = repository.getCountries();
        }
    }

    public void getLiveProvinceData(String value){
        if (!value.equals(currentVal) || provinceData.getValue() == null) {
            currentVal = value;
            currentState.setValue(LOADING);
            provinceData = repository.getProvince(value);

            }
    }

}

File: app/src/main/java/com/example/homeworkproject/viewmodels/LocationViewModel.kt
package com.example.homeworkproject.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.MutableLiveData
import com.example.homeworkproject.model.ApiState
import com.example.homeworkproject.model.Province
import com.example.homeworkproject.model.Country
import com.example.homeworkproject.repository.Repository
import java.util.ArrayList

class LocationViewModel : ViewModel() {
    private val repository: Repository = Repository()

    @JvmField
    val currentState = MutableLiveData<ApiState.Status>()
    var provinceData: MutableLiveData<ArrayList<Province>>? = null
    var countryData: MutableLiveData<ArrayList<Country>>? = null
    private var currentVal: String? = null

    fun getLiveCountryData() {
            if (countryData == null || countryData!!.value == null) {
                currentState.value = ApiState.Status.LOADING
                countryData = repository.getCountries()
            }
        }

    fun getLiveProvinceData(value: String?) {
        if (value != currentVal || provinceData!!.value == null) {
            currentVal = value
            currentState.value = ApiState.Status.LOADING
            provinceData = repository.getProvinces(value)
        }
    }

}
